agent/lib_composer.c

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright 2022 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "php_agent.h"
+#include "fw_hooks.h"
+#include "fw_support.h"
+#include "nr_txn.h"
+#include "util_logging.h"
+#include "util_memory.h"
+#include "util_syscalls.h"
+
+static bool nr_execute_handle_autoload_composer_is_initialized() {
+  zend_class_entry* zce = NULL;
+
+  if (NULL == (zce = nr_php_find_class("composer\\installedversions"))) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "Composer\\InstalledVersions class not found");
+    return false;
+  };
+
+  // the class is found - there's hope!
+  if (NULL == nr_php_find_class_method(zce, "getallrawdata")
+      || NULL == nr_php_find_class_method(zce, "getrootpackage")) {
+    nrl_verbosedebug(
+        NRL_INSTRUMENT,
+        "Composer\\InstalledVersions class found, but methods not found");
+    return false;
+  }
+
+  return true;
+}
+
+static int nr_execute_handle_autoload_composer_init(const char* vendor_path) {
+  char* code = NULL;
+  zval retval;
+  int result = -1;
+
+  if (nr_execute_handle_autoload_composer_is_initialized()) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: already initialized", __func__);
+    return NR_SUCCESS;
+  }
+
+  code = nr_formatf("include_once '%s/composer/InstalledVersions.php';",
+                    vendor_path);
+
+  result = zend_eval_string(code, &retval, "newrelic\\init_composer_api");
+  if (result != SUCCESS) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: zend_eval_string(%s) failed, result=%d", __func__,
+                     code, result);
+    nr_free(code);
+    return NR_FAILURE;
+  }
+
+  zval_dtor(&retval);
+  nr_free(code);
+
+  // Make sure runtime API is available after loading
+  // Composer\\InstalledVersions class:
+  if (!nr_execute_handle_autoload_composer_is_initialized()) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: unable to initialize Composer runtime API", __func__);
+    return NR_FAILURE;
+  }
+
+  return NR_SUCCESS;
+}
+
+static void nr_execute_handle_autoload_composer_get_packages_information(
+    const char* vendor_path) {
+  zval retval;  // This is used as a return value for zend_eval_string.
+                // It will only be set if the result of the eval is SUCCESS.
+  int result = -1;
```

Comment:
```suggestion
  int result = FAILURE;
```

You are comparing it to a [zend_result](https://github.com/php/php-src/blob/cf0a44a38a512e238db490282adf80682200d296/Zend/zend_types.h#L59) later, so we should ensure we stay in sync with the zend_result values.